File: C++/Step17-Divide and Conquer/1992.cpp
//
//  1992.cpp
//  Step17-Divide and Conquer
//
//  Created by 김예빈 on 2019. 9. 20..
//  Copyright © 2019년 김예빈. All rights reserved.
//

#include <iostream>
using namespace std;

int square[64][64];

void divide(int, int, int);

int main(int argc, const char * argv[]) {
    int n;
    scanf("%d", &n);
    
    for(int i = 0; i < n; i++)
        for(int j = 0; j < n; j++)
            scanf("%1d", &square[i][j]);
    
    divide(0, 0, n);
    
    return 0;
}

void divide(int x, int y, int size) {
    int color = square[y][x];
    bool check = true;
    
    if(size == 1) {
        printf("%d", color);
        return;
    }
    
    for(int i = y; i < y + size; i++) {
        if(!check) break;
        for(int j = x; j < x + size; j++) {
            if(color != square[i][j]) {
                check = false;
                break;
            }
        }
    }
    
    if(check)
        printf("%d", color);
    else {
        printf("(");
        divide(x, y, size / 2);
        divide(x + size / 2, y, size / 2);
        divide(x, y + size / 2, size / 2);
        divide(x + size / 2, y + size / 2, size / 2);
        printf(")");
    }
}

File: BaekJoon Online Judge/Swift/Step25-Divide and Conquer/1992.swift
//
//  1992.swift
//  Step17-Divide and Conquer
//
//  Created by Yebin Kim on 2020/01/20.
//  Copyright © 2020 김예빈. All rights reserved.
//

func divide(_ x: Int, _ y: Int, size: Int, image: [[Int]]) {
    let color = image[y][x]
    var check = true
    
    for i in y..<(y + size) {
        if !check { break }
        
        for j in x..<(x + size) {
            if color != image[i][j] {
                check = false
                break
            }
        }
    }
    
    if check {
        print(color, terminator: "")
    } else {
        print("(", terminator: "")
        divide(x, y, size: size / 2, image: image)
        divide(x + size / 2, y, size: size / 2, image: image)
        divide(x, y + size / 2, size: size / 2, image: image)
        divide(x + size / 2, y + size / 2, size: size / 2, image: image)
        print(")", terminator: "")
    }
}

var image = [[Int]]()

for _ in 0..<Int(readLine()!)! {
    image.append(readLine()!.compactMap({ Int(String($0))! }))
}

if image.count <= 1 {
    print("(\(image.first!.first!))")
} else {
    divide(0, 0, size: image.count, image: image)
}
